AwsCryptographyPrimitives/runtimes/python/src/aws_cryptography_primitives/internaldafny/extern/ECDH.py

```diff
@@ -0,0 +1,394 @@
+        
+        curve = maybe_ecc_algorithm.value
+        
+        if curve.name != "SM2PKE": # magic string
+            try:
+                private_key = load_pem_private_key(private_key_pem_bytes, None)
+                public_key = load_der_public_key(public_key_der_bytes)
+
+                shared_key = private_key.exchange(
+                    ec.ECDH(),
+                    public_key
+                )
+
+                return CreateExternDerivesharedSecretSuccess(
+                    _dafny.Seq(
+                        shared_key
+                    )
+                )
+            except Exception as e:
+                return CreateExternDerivesharedSecretError(
+                    _smithy_error_to_dafny_error(
+                        AwsCryptographicPrimitivesError(
+                            message=str(e)
+                        )
+                    )
+                )
+        else:
+            return CreateExternDerivesharedSecretError(
+                _smithy_error_to_dafny_error(
+                    ValueError(
+                        "SM2 not yet supported."
+                    )
+                )
+            )
+
+class KeyGeneration:
+    
+    @staticmethod
+    def GenerateKeyPair(dafny_eccAlgorithm) -> Wrappers.Result:
+        maybe_ecc_algorithm = ECCAlgorithms.eccAlgorithm(dafny_eccAlgorithm)
+        if maybe_ecc_algorithm.is_Failure:
+            return CreateExternEccKeyGenFailure(
+                maybe_ecc_algorithm.error
+            )
+        
+        curve = maybe_ecc_algorithm.value
+
+        if curve.name != "SM2PKE":
+            try:
+                private_key = ec.generate_private_key(
+                    maybe_ecc_algorithm.value.value
+                )
+
+                private_key_pem_bytes = private_key.private_bytes(Encoding.PEM, PrivateFormat.PKCS8, NoEncryption())
+                public_key_der_bytes = private_key.public_key().public_bytes(Encoding.DER, PublicFormat.SubjectPublicKeyInfo)
+
+                return CreateExternEccKeyGenSuccess(EccKeyPair_EccKeyPair(
+                    publicKey=_dafny.Seq(public_key_der_bytes),
+                    privateKey=_dafny.Seq(private_key_pem_bytes)
+                ))
+            except Exception as e:
+                return CreateExternEccKeyGenFailure(
+                    _smithy_error_to_dafny_error(
+                        AwsCryptographicPrimitivesError(
+                            message=str(e)
+                        )
+                    )
+                )
+        else:
+            return CreateExternEccKeyGenFailure(
+                    _smithy_error_to_dafny_error(
+                        ValueError(
+                            "SM2 not supported."
+                        )
+                    )
+                )
+
+
+class ECCUtils:
+
+    CURVE_TO_ECC_SECRET_LENGTH_MAP = {
+        "secp256r1": int(256 / 8),
+        "secp384r1": int(384 / 8),
+        "secp521r1": int(521 / 8 + 1)
```

Comment:
in the java code we wrote a comment about the magic behind the +1, let's add that here too